include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -288,6 +288,77 @@ inline char* Session::EndProfiling(OrtAllocator* allocator) const {
   return out;
 }
 
+inline std::vector<Value> ExperimentalSession::Run(const Value& input_values, const RunOptions& run_options) {
+  size_t output_node_count = GetOutputNames().size();
```

Comment:
Not sure this adds enough value now given I could do the following couldn't I?

`Run(GetInputNames(), input_values, ...)`
`Run(GetInputNames(), input_values, GetOutputNames(), output_values, ...)

Reply:
The value in this particular overloaded function (in my opinion) is the coder doesn't need to know anything about the input or output layer names in order to get started running inference. This design choice/pattern shows up in other frameworks as well (PyTorch and Keras for example).

The idea is that if `input_values` is the only supplied argument, then the API will _smartly_ assume that all model inputs and outputs are used. Ownership is on the coder to ensure their input data has the correct shape and check output.

I tried to make just two overloaded `Run()` methods by defining default values for `input_names` and `output_names` but the code was not clean - those parameters cannot be set with `GetInputNames()` and `GetOutputNames()` as  they are member functions and the working solution was too verbose in my opinion. Having extra overloaded `Run()` seemed cleaner.

The way to call `Run()` with your suggestion would be
```
session.Run(input_tensors, session.GetInputNames(), session.GetOutputNames());
```
which shows the extra verbosity I'm referring to.

Reply:
Some additional info in case it helps:
  - Valid inputs when running the model are the union of GetInputNames (required inputs) and GetOverridableInitializerNames (optional inputs). One example usage is a model that processes a sequence in multiple chunks, where the state that was output from the previous call to Run is passed in as an optional input when continuing. 
  - You can request a subset of outputs to avoid the cost of returning something you don't need

Coming from the perspective that we want to keep the API as simple as possible, I can see value in adding the Get*Names and Get*Shapes methods to the Session class, and I can see value in adding a general purpose Run(...) like this:
```
// provide an empty vector for output_values unless you have preallocated Value instances for the outputs to be written to
void Run(const std::vector<std::string>& input_names,
         const std::vector<Value>& input_values, 
         const std::vector<std::string>& output_names,
         std::vector<Value>& output_values,
         const RunOptions& run_options = RunOptions());
```

The reason for explicitly specifying the names is that the order must match the entries in the input_values and output_values vectors. For the core ORT C++ API it feels a little too opaque to have Run methods where this relationship is not clear from the arguments to the call. That potentially leads to confusion from users and github Issues that need to be responded to. It also makes things like error messages easier to understand for the user. e.g. say the first input was an invalid type, and I get an error message saying "Incorrect type for input 'input_name'", that's much easier for the user to understand when they provided a list of input names and matching input values vs just the input values as the latter requires an implicit knowledge of which entry in input_values that name applies to.

Putting together the vector of input and output names once and passing it in to Run on each call makes the cost of the Get*Names calls where a buffer is allocated, copied to a std::string, and deallocated for each input/output a one-off. It also supports scenarios involving optional inputs and a subset of outputs. It should be trivial to add a wrapper to this Run method if say you want to avoid passing in the input names. 

Should it support something other than std::vector via templatization? Given the containers need to be ordered that would allow using it with something like std::array or std::deque, however comes at the potential cost that a new user might do something wrong and get a confusing compiler error about invalid template usage, and create a github Issue that needs to be responded to. I would suggest it's better to prove there are scenarios where std::vector is insufficient before supporting more container types via templates.

Reply:
@skottmckay thanks for the explanation! I wasn't very clear on what the labels were for since my own model has one "unlabled" big input array (which I just label "input" because it needs a label), and one simple output. 
As the API is, the labels being disconnected from the data makes their purpose harder to see for n00bs to the API like me. Spitballing/going off on a tangent... if being able to re-order inputs is useful and desired, then maybe they would be better implemented as separate input vectors inside a container (e.g. vector of {label, isOptional, type, size, data}). I am imagining this container structure as a base class, to which each project defines their own implementation that defines the input/output labels and structure. Then it is very explicit when setting up the input to Run which part of the input has what label, size, etc... half-baked thought.

Coming back from my tangent: I see now from @skottmckay's explanation why his Run declaration makes the most sense right now. One day I'd love to see a code path that optimizes for the case with no optional inputs/outputs and forced ordering that would use the labels stored in the session instead of needing to pass them in.